README.pod
=pod

=head1 NAME

OAuth2::Provider - A generic OAuth2 Provider

=head1 VERSION

version 0.1

=head1 SYNOPSIS

OAuth2 clients are plenty but provider are harder to find than they
should be. This provider is written in a way that hopefully makes it
easy to tweak for your environment.

=head1 ATTRIBUTES

=head2 attr

=head1 SUBROUTINES/METHODS

General authentocator implementing all authentication mechanisms

=head2 authenticator

=head2 get_client_by_id

Get a client definition by client_id

=head2 validate_by_token

Validate a token, check it for expiration and return the associated
info for the token. This accepts a optional second parameter if the
requests is for a refresh_token.

=head2 validate_user_by_username

Validate a user/password combination

=head2 get_access_token

Get a access_token from the storage plugin

=head2 get_refresh_token

Get a AccessToken by token

=head2 create_access_token

Create access token based on the client information. This has a optional
second parameter if the token are based off a refresh_token

=head2 to_bearer_token

Convert a token to a Bearer Token

=head2 to_mac_token

Convert a token to a MAC token

=head2 error

General error handler

=head1 BUGS

Please report any bugs or feature requests on GitHub's issue tracker L<https://github.com/<github_user>/OAuth2::Provider/issues>.
Pull requests welcome.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc OAuth2::Provider

You can also look for information at:

=over 4

=item * GitHub repository

L<https://github.com/<github_user>/OAuth2::Provider>

=item * MetaCPAN

L<https://metacpan.org/module/OAuth2::Provider>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/OAuth2::Provider>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/OAuth2::Provider>

=back

=head1 ACKNOWLEDGEMENTS

=over 4

=item * Lenz Gschwendtner (@norbu09), for being an awesome mentor and friend.

=back

=head1 AUTHOR

Lenz Gschwendtner <lenz@springtimesoft.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by springtimesoft LTD.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.



OAuth2-Provider-0.1/Changes
==================================================
Changes from 2012-06-13 00:00:00 +0000 to present.
==================================================

----------------------------------------
version 0.1 at 2013-06-13 04:10:44 +0000
----------------------------------------

  Change: 659e298abab0abf119685406e84f6199a5e42610
  Author: Lenz Gschwendtner <lenz@springtimesoft.com>
  Date : 2013-06-13 16:07:25 +0000

    first few lines of a general purpos OAuth2 provider base class 

  Change: 4cab7b0841509486ecfba5e11e6154dd1d53df5a
  Author: Lenz Gschwendtner <lenz@springtimesoft.com>
  Date : 2013-06-13 14:30:21 +0000

    initial commit 

================
End of releases.
================


lib/OAuth2/Provider.pm
package OAuth2::Provider;

use 5.012;
use Any::Moose;

# ABSTRACT: A generic OAuth2 Provider

# VERSION


=head1 SYNOPSIS

OAuth2 clients are plenty but provider are harder to find than they
should be. This provider is written in a way that hopefully makes it
easy to tweak for your environment.

=head1 ATTRIBUTES

=head2 attr

=cut

has 'attr' => (
    is => 'rw',
);



=head1 SUBROUTINES/METHODS

General authentocator implementing all authentication mechanisms

=head2 authenticator

=cut

sub authenticator { }


=head2 get_client_by_id

Get a client definition by client_id

=cut

sub get_client_by_id { }



=head2 validate_by_token

Validate a token, check it for expiration and return the associated
info for the token. This accepts a optional second parameter if the
requests is for a refresh_token.

=cut

sub validate_by_token { }


=head2 validate_user_by_username

Validate a user/password combination

=cut

sub validate_user_by_username { }


=head2 get_access_token

Get a access_token from the storage plugin

=cut

sub get_access_token { }

=head2 get_refresh_token

Get a AccessToken by token

=cut

sub get_refresh_token { }

=head2 create_access_token

Create access token based on the client information. This has a optional
second parameter if the token are based off a refresh_token

=cut

sub create_access_token {
    #body ...
}


=head2 to_bearer_token

Convert a token to a Bearer Token

=cut

sub to_bearer_token { }


=head2 to_mac_token

Convert a token to a MAC token

=cut

sub to_mac_token { }

=head2 error

General error handler

=cut

sub error {
    my($self, $error) = @_;

    warn $error;
    return;
}


=head1 BUGS

Please report any bugs or feature requests on GitHub's issue tracker L<https://github.com/<github_user>/OAuth2::Provider/issues>.
Pull requests welcome.


=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc OAuth2::Provider


You can also look for information at:

=over 4

=item * GitHub repository

L<https://github.com/<github_user>/OAuth2::Provider>

=item * MetaCPAN

L<https://metacpan.org/module/OAuth2::Provider>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/OAuth2::Provider>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/OAuth2::Provider>

=back


=head1 ACKNOWLEDGEMENTS

=over 4

=item * Lenz Gschwendtner (@norbu09), for being an awesome mentor and friend.

=back

=cut

1;  # End of OAuth2::Provider


lib/OAuth2/Provider/Plugins/CouchDB.pm
package OAuth2::Provider::Plugin::CouchDB;

use 5.012;
use Any::Moose 'Role';
use Store::CouchDB;

# ABSTRACT: CouchDB Storage provider
#
# Version

=head1 ATTRIBUTES

=head2 couch_host

=cut

has 'couch_host' => (
    is      => 'rw',
    isa     => 'Str',
    lazy    => 1,
    default => sub { 'localhost' },
);

=head2 couch_port

=cut

has 'couch_port' => (
    is      => 'rw',
    isa     => 'Int',
    lazy    => 1,
    default => sub { 5984 },
);

=head2 couch_db

=cut

has 'couch_db' => (
    is      => 'rw',
    isa     => 'Str',
    lazy    => 1,
    default => sub { 'test' },
);

=head2 couch_user

=cut

has 'couch_user' => (
    is  => 'rw',
    isa => 'Str',
);

=head2 couch_pass

=cut

has 'couch_pass' => (
    is  => 'rw',
    isa => 'Str',
);

has 'couchdb' => (
    is  => 'rw',
    isa => 'Store::CouchDB',
    builder  => '_connect_couchdb',
);

sub _connect_couchdb {
    my $self = shift;

    my $couch = Store::CouchDB->new(
        host  => $self->couch_host,
        port  => $self->couch_port,
        debug => $self->debug ? 1 : 0,
    );
    $couch->db($self->couch_db) if ($self->couch_db);
    if ($self->couch_user && $self->couch_pass) {
        $couch->user($self->couch_user);
        $couch->pass($self->couch_pass);
    }
    return $couch;
}


1;


lib/OAuth2/Provider/Plugins/KyotoTycoon.pm
package OAuth2::Provider::Plugin::KyotoTycoon;

use 5.012;
use Any::Moose 'Role';
use Cache::KyotoTycoon;
use Try::Tiny;
use MIME::Base64 qw(encode_base64 decode_base64);
use Storable qw/nfreeze thaw/;

# ABSTRACT: KyotoTycoon Storage provider
#
# Version

=head1 ATTRIBUTES

=head2 tycoon_host

=cut

has 'tycoon_host' => (
    is      => 'rw',
    isa     => 'Str',
    lazy    => 1,
    default => sub { 'localhost' },
);

=head2 tycoon_port

=cut

has 'tycoon_port' => (
    is      => 'rw',
    isa     => 'Int',
    lazy    => 1,
    default => sub { 1978 },
);

=head2 tycoon_db

=cut

has 'tycoon_db' => (
    is      => 'rw',
    isa     => 'Str',
    lazy    => 1,
    default => sub { 0 },
);

=head2 tycoon_connect_timeout

=cut

has 'tycoon_timeout' => (
    is      => 'rw',
    isa     => 'Int',
    lazy    => 1,
    default => sub { 5 },
);

=head2 tycoon

=cut

has 'tycoon' => (
    is       => 'rw',
    isa      => 'Cache::KyotoTycoon',
    builder  => '_connect_tycoon',
);

sub _connect_tycoon {
    my $self = shift;

    my $tycoon = KyotoTycoon->new(
        host    => $self->tycoon_host,
        port    => $self->tycoon_port,
        timeout => $self->tycoon_timeout,
        db      => $self->tycoon_db,
    );
    return $tycoon;
}

=head2

cache_get is used to get a value from a fast cache as opposed to a slow database. this should be used to look up access_token

=cut

sub cache_get {
    my ($self, $key) = @_;

    my ($value, $expire) = $self->tycoon->get($key);
    return unless defined $value;
    return (thaw(decode_base64($value)), $expire);
}

=head2

cache_set is used to set a value in a fast cache as opposed to a slow database. this should be used to store access_token

=cut

sub cache_set {
    my ($self, $key, $data, $expire) = @_;

    $data = encode_base64(nfreeze($data));
    try {
        return $self->tycoon->replace($key, $data, $expire);
    }
    catch {
        return $self->tycoon->set($key, $data, $expire)
    };
    return 1;
}

1;


lib/OAuth2/Provider/Plugins/Token/AuthToken.pm
package OAuth2::Provider::Plugin::Token;

use 5.012;
use Any::Moose 'Role';
use Cache::KyotoTycoon;
use Try::Tiny;
use MIME::Base64 qw(encode_base64 decode_base64);
use Storable qw/nfreeze thaw/;

# ABSTRACT: AuthToken Storage provider backed by KyotoTycoon
#
# Version



around 'validate_by_token' => sub {
    my ($orig, $self, $token, $client) = @_;

    if($client){
        return $self->get_refresh_token($token);
    } else {
        return $self->get_access_token($token);
    }
};

around 'get_access_token' => sub {
    my ($orig, $self, $token) = @_;

    return $self->cache_get($token);
};

around 'get_refresh_token' => sub {
    my ($orig, $self, $token) = @_;

    # find refresh token in couchdb
};

around 'create_access_token' => sub {
    my ($orig, $self, $client, $refresh_token) = @_;

    # generate a access token, save it in the cache, log it somewhere.
    # generate a refresh token and save it in the DB
};

1;


OAuth2-Provider-0.1/lib/OAuth2/Provider/Plugins/Token.pm
package OAuth2::Provider::Plugin::Token;

use 5.012;
use Any::Moose 'Role';
use Data::UUID;

# ABSTRACT: Token provider for OAuth2 Provider
#
# Version


has 'token_expire' => (
    is       => 'rw',
    isa      => 'Int',
    required => 1,
    default  => 3600,
);


has 'token' => {
    is => 'rw',
    isa => 'Str',
    lazy => 1,
    builder => 'generate_token'
};


around 'authenticator' => sub {
    my ($orig, $self, $req, $mode) = @_;

    my $client = $self->get_client_by_id($req->{client_id});
    return $self->error('client_not_found') unless $client;

    return $self->error('client_authentication_failed')
        unless $client->{client_secret} eq $req->{client_secret};

    my $token;
    given ($req->{grant_type}) {
        when (/^authorization_code$/i) {
            my $code = $self->validate_by_token($req->{token});
            return $self->error('no_valid_token') unless $code;

            # TODO is this where we also need to add support for roles?
            return $self->error('invalid_grant')
                unless $code->{redirect_uri} eq $req->{redirect_uri};
            $token = $self->get_access_token($code);
        }
        when (/^password$/i) {
            my $user =
                $self->validate_user_by_username($req->{username},
                $req->{password});
            return $self->error('invalid_grant') unless $user;
            $token = $self->create_access_token($client);
        }
        when (/^client_credentials$/i) {

            # already authenticated
            my $token = $self->create_access_token($client);
            return { access_token => $token };
        }
        when (/^refresh_token$/i) {
            my $code = $self->validate_by_token($req->{refresh_token}, $client);
            return $self->error('no_valid_token') unless $code;
            $token = $self->create_access_token($client, $code);
        }
        default { return $self->error('unsupported_grant_type'); }
    }
    given ($mode) {
        when ('mac') {
            return $self->to_mac_token({ access_token => $token });
        }
        default {
            return $self->to_bearer_token({ access_token => $token });
        }
    }
};

sub generate_token {
    my $self = shift;

    my $ug = new Data::UUID;
    my $token = $ug->create_b64();
    $token = s/\W//g;
    return $token;
}

1;

__END__

=pod

=head1 NAME

OAuth2::Provider::Plugin::Token - Token provider for OAuth2 Provider

=head1 VERSION

version 0.1

=head2 token_expire

Default token expire in seconds.

=head2 token

Implicitly generate a token

=head1 AUTHOR

Lenz Gschwendtner <lenz@springtimesoft.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by springtimesoft LTD.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


OAuth2-Provider-0.1/Makefile.PL

use strict;
use warnings;

use 5.012;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "A generic OAuth2 Provider",
  "AUTHOR" => "Lenz Gschwendtner <lenz\@springtimesoft.com>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "OAuth2-Provider",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "OAuth2::Provider",
  "PREREQ_PM" => {
    "Any::Moose" => 0,
    "Data::UUID" => 0
  },
  "TEST_REQUIRES" => {},
  "VERSION" => "0.1",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



